commands/root.go

```diff
@@ -16,55 +16,131 @@
 	mapcmd "github.com/hazelcast/hazelcast-commandline-client/commands/types/map"
 	"github.com/hazelcast/hazelcast-commandline-client/internal"
+	"github.com/hazelcast/hazelcast-commandline-client/internal/cobraprompt"
 )
 
 var (
-	cfgFile   string
-	addresses string
-	cluster   string
-	token     string
-	rootCmd   = &cobra.Command{
+	RootCmd = &cobra.Command{
 		Use:   "hzc {cluster | help | map} [--address address | --cloud-token token | --cluster-name name | --config config]",
 		Short: "Hazelcast command-line client",
 		Long:  "Hazelcast command-line client connects your command-line to a Hazelcast cluster",
 		Example: "`hzc map --name my-map put --key hello --value world` - put entry into map directly\n" +
 			"`hzc help` - print help",
 		RunE: func(cmd *cobra.Command, args []string) error {
+			if _, err := internal.MakeConfig(); err != nil {
+				return err
+			}
 			return cmd.Help()
 		},
 	}
 )
 
+func addressAndClusterNamePrefix() (prefix string, useLive bool) {
+	return fmt.Sprintf("hzc %s@%s> ", internal.Configuration.Cluster.Network.Addresses[0], internal.Configuration.Cluster.Name), true
+}
+
+var advancedPrompt = &cobraprompt.CobraPrompt{
+	RootCmd:                  RootCmd,
+	PersistFlagValues:        true,
+	ShowHelpCommandAndFlags:  true,
+	ShowHiddenFlags:          true,
+	SuggestFlagsWithoutDash:  true,
+	DisableCompletionCommand: true,
+	AddDefaultExitCommand:    true,
+	GoPromptOptions: []prompt.Option{
+		prompt.OptionTitle("hazelcast commandline client"),
+		prompt.OptionLivePrefix(addressAndClusterNamePrefix),
+		prompt.OptionMaxSuggestion(10),
+	},
+	OnErrorFunc: func(err error) {
+		// handle error noop to prevent application from crashing
+		return
+	},
+}
+
+func IsInteractiveCall() bool {
+	cmd, flags, err := RootCmd.Find(os.Args[1:])
+	if err != nil {
+		return false
+	}
+	for _, flag := range flags {
+		if flag == "--help" || flag == "-h" {
+			return false
+		}
+	}
+	if cmd.Name() == "help" {
+		return false
+	}
+	if cmd == RootCmd {
+		return true
+	}
+	return false
+}
+
 func Execute() {
-	if err := rootCmd.Execute(); err != nil {
+	if err := RootCmd.Execute(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func ExecuteInteractive() {
+	// parse global persistent flags
+	if err := RootCmd.ParseFlags(os.Args); err != nil {
 		log.Fatal(err)
 	}
+	// initialize global config
+	conf, err := internal.MakeConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println("Connecting to the cluster ...")
+	if _, err = internal.ConnectToCluster(context.Background(), conf.Clone()); err != nil {
+		fmt.Printf("Error: %s\n", err)
+		return
+	}
+	var flagsToExclude []string
+	RootCmd.PersistentFlags().VisitAll(func(flag *pflag.Flag) {
+		flagsToExclude = append(flagsToExclude, flag.Name)
+	})
+	flagsToExclude = append(flagsToExclude, "help")
+	advancedPrompt.FlagsToExclude = flagsToExclude
+	RootCmd.Example = "> map put -k key -m myMap -v someValue\n" +
+		"> map get -k key -m myMap\n" +
+		"> cluster version"
```

Comment:
Instead of concatenating strings and adding line breaks, backticks (````) may be used.

Reply:
Great catch, thanks